Bug 1517700 - Add a fuzzy job finder to Treeherder

KWierso opened the pull request.

edmorley commented on package.json:
```diff
@@ -29,6 +29,7 @@
     "bootstrap": "4.2.1",
     "d3": "5.7.0",
     "font-awesome": "4.7.0",
+    "fuse.js": "3.3.0",
```
This library seems pretty neat:
http://fusejs.io/

...and is small enough that it should be fine to use:
https://bundlephobia.com/result?p=fuse.js@3.3.0

@camd, I wonder if this library could be used to simplify the jobs-view filter implementation at a later point? :-)

camd replied:
Yeah, looks interesting.  I'll check it out.  :)

edmorley commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,293 @@
+/* eslint react/no-multi-comp: 0, react/prop-types: 0 */
```
(These eslint-disable lines will need to be removed/fixed before final review)

KWierso replied:
Done.

edmorley commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,293 @@
+            </label>
+            <InputGroup id="addJobsGroup">
+              <select
+                ref={this.addJobsRef}
```
Controlled components (https://reactjs.org/docs/forms.html#controlled-components) are normally preferred over uncontrolled components (https://reactjs.org/docs/uncontrolled-components.html). It might be worth switching this to the former, unless it doesn't work well with fuse.js for some reason?

KWierso replied:
I think I fixed this. It works without the refs, at least.

edmorley commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,293 @@
+  };
+
+  // Implement an alphanumeric sort so mochitest-10 is sorted after mochitest 9, not mochitest-1
+  sortAlphaNum = (a, b) => {
```
I think it might be worth moving this under `ui/helpers/`

KWierso replied:
Done.

camd commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,276 @@
+    super(props);
+
+    this.state = {
+      modal: this.props.modal,
```
Do we need ``modal`` in the state?  We already have it in the ``props``?

camd replied:
Actually, maybe we don't need ``modal`` as a prop either.  I don't think it provides anything useful, does it?  I might be missing something, though.  :)

KWierso replied:
I think I copied it from some example somewhere and misused it. Gone, and everything seems to work still.

camd commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,276 @@
+    }
+  };
+
+  toggle = () => {
```
I don't think this function is used.

camd commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,276 @@
+          toggle={this.props.toggle}
+          className={this.props.className}
+        >
+          <ModalHeader toggle={this.state.toggle}>
```
I don't think there is a ``toggle`` in ``state``

KWierso commented on ui/css/treeherder-fuzzyfinder.css:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Fuzzy runnable job finder
+ */
+.fuzzy-modal select {
```
Done. Here's the new look:

![screenshot_2019-01-11 99 - 6 try fuzzy query web-platform](https://user-images.githubusercontent.com/172215/51027066-d068cf80-1544-11e9-8b52-ef3d803b6d14.png)

KWierso commented:
Comments addressed.

Before this lands, I'd really like to trim down the list of runnable jobs to hopefully avoid making it easy to accidentally trigger 11000+ jobs, including some jobs that should definitely not be run by mere mortals.

I could either trim the list down and prevent the other jobs from ever being included, or I could add a toggle to switch between the trimmed down list and the full, dangerous list.

For comparison, `mach try fuzzy` on the command line by default only lists 2000 or so jobs, but the other 9000 or so can be run if you do `mach try fuzzy --full`. In Treeherder right now with this PR, the 11000 jobs are listed by default, and the "Add all" button looks very pushable. I'd like to switch things so I trim down the list to 2000 or so jobs before popping up the modal, and maybe add a checkbox that could let the other jobs be included in the results.

camd commented:
> Comments addressed.
> 
> Before this lands, I'd really like to trim down the list of runnable jobs to hopefully avoid making it easy to accidentally trigger 11000+ jobs, including some jobs that should definitely not be run by mere mortals.
> 
> I could either trim the list down and prevent the other jobs from ever being included, or I could add a toggle to switch between the trimmed down list and the full, dangerous list.
> 
> For comparison, `mach try fuzzy` on the command line by default only lists 2000 or so jobs, but the other 9000 or so can be run if you do `mach try fuzzy --full`. In Treeherder right now with this PR, the 11000 jobs are listed by default, and the "Add all" button looks very pushable. I'd like to switch things so I trim down the list to 2000 or so jobs before popping up the modal, and maybe add a checkbox that could let the other jobs be included in the results.

Yeah, this makes good sense.  It's really slow to load right now, just due to the size of the list.  Having it start with a pared down list would be great.  Having a checkbox to "show all" sounds like a great idea.  Not sure if there's a way to show "common" tests at first, and the show all adds the rest.  Or if you'd just have to show the first 2000 or something.  I'll leave that up to your expertise.  :)

For making the lists faster, you may check out ``react-virtualized``: https://github.com/bvaughn/react-virtualized

It's supposed to speed up rendering of large lists a lot.

camd commented:
The XHR query is only 542ms for me.  I think rendering the items in the list is where most of the time is spent.  So if you only rendered 2000, it'll probably render much quicker.

camd commented:
I should have said: "So yeah, I agree with your approach."  :)

edmorley commented on ui/css/treeherder-fuzzyfinder.css:
```diff
@@ -0,0 +1,8 @@
+ * Fuzzy runnable job finder
+ */
+
+.fuzzy-modal .modal-body #addJobsGroup option.selected::before {
```
Now that #4489 has landed, this will need a rebase and then a `yarn format` so the CSS style in this newly added file is updated :-)

KWierso commented:
Okay, it now uses a vastly pared-down list of jobs by default (~3000 vs ~12000), with an option to search the full list included. The pared down list should be all of the jobs that could run by default on pushes to mozilla-central, and shouldn't include any esoteric stuff like release-promotion stuff or toolchain jobs. This shorter list renders much more quickly than the full list, which is nice.

KWierso commented on ui/job-view/pushes/Push.jsx:
```diff
@@ -337,6 +339,45 @@ class Push extends React.Component {
 
+  showFuzzyJobs = async () => {
+    const { push, repoName, getGeckoDecisionTaskId, notify } = this.props;
+    const excludedJobNames = /(balrog|beetmover|bouncer-locations-firefox|build-docker-image|build-(.+)-nightly|build-(.+)-upload-symbols|checksums|cron-bouncer|dmd|fetch|google-play-strings|push-to-release|mar-signing|nightly|packages|release-bouncer|release-early|release-final|release-secondary|release-snap|release-source|release-update|repackage-l10n|repo-update|searchfox|sign-and-push|test-(. ...
```
@edmorley @camd Not super happy with that super-long regex for trimming the list down, but I don't know enough regex-fu to make it any shorter. Any chance either of you know if I could trim that down any, or maybe break it onto multiple lines so it's a bit easier to read? 

I basically just want any job name that contains any of those strings to be removed.

KWierso replied:
[Tagged Templates](https://stackoverflow.com/questions/12317049/how-to-split-a-long-regular-expression-into-multiple-lines-in-javascript#answers) let me break this into multiple lines.

KWierso commented:
Made a few minor UI tweaks and rebased to a current master branch, and fixed up some recently enabled eslint rules.

Here's the (hopefully) final appearance:

![Screenshot_2019-03-26  3  mozilla-inbound](https://user-images.githubusercontent.com/172215/55037663-b2efc180-4fdb-11e9-935d-a4279af6ebfe.png)

camd commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    if (ev && ev.type === 'keydown') {
+      if (ev.keyCode === 13) {
```
Use of ``keyCode`` is now deprecated.  We should use ``key`` instead: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent

KWierso replied:
Done. Looks like there's a couple other uses of keyCode in the repo that should get fixed at some point: https://github.com/mozilla/treeherder/search?q=keyCode&unscoped_q=keyCode

camd replied:
Oh, heh.  I guess I missed those when I converted it.  My bad!  Travis has been a little wonky today.  But I'll keep an eye on this to see when the tests pass...

camd commented on ui/job-view/pushes/FuzzyJobFinder.jsx:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    if (ev && ev.type === 'keydown') {
+      if (ev.key === "Enter") {
```
This needs single quotes, instead of double.
